lib/DataFetcher/DataFetcher.js

```diff
@@ -0,0 +1,118 @@
+import React, {
+  useState,
+  useEffect,
+  useRef,
+} from 'react';
+import PropTypes from 'prop-types';
+import { get } from 'lodash';
+
+import { DEFAULT_FETCHER_UPDATE_INTERVAL } from '../utils/constants';
+
+export function DataFetcher({
+  FetcherContext,
+  updateInterval,
+  mutator,
+  resources,
+  resourcesMappingPath,
+  children,
+}) {
+  const mapResourcesToState = ({
+    hasLoaded,
+    isEmpty,
+  }) => {
+    const newContextData = { hasLoaded };
+
+    Object.keys(resources).forEach(key => {
+      newContextData[key] = isEmpty ? [] : get(resources[key], resourcesMappingPath, []);
+    });
+
+    return newContextData;
+  };
+
+  const [contextData, setContextData] = useState(() => mapResourcesToState({
+    hasLoaded: false,
+    isEmpty: true,
+  }));
+
+  const timeoutId = useRef(null);
+
+  const resetFetcher = () => {
+    clearTimeout(timeoutId.current);
+    timeoutId.current = null;
+  };
+
+  const fetchResources = async initial => {
+    const resourcesPromises = Object
+      .values(mutator)
+      .reduce((res, resourceMutator) => res.concat(fetchResource(resourceMutator)), []);
+
+    try {
+      await Promise.all(resourcesPromises);
+
+      setContextData(mapResourcesToState({
+        isEmpty: false,
+        hasLoaded: true,
+      }));
+    } catch (error) {
+      if (initial) {
+        // fill contextData with empty data on unsuccessful initial data retrieval
+        setContextData(mapResourcesToState({
+          isEmpty: true,
+          hasLoaded: true,
+        }));
+      }
+    }
+  };
+
+  useEffect(() => resetFetcher, []);
+
+  useEffect(() => {
+    function updateResourcesData(initial) {
+      if (!initial && timeoutId.current === null) return;
+
+      resetFetcher();
+
+      timeoutId.current = setTimeout(async () => {
+        await fetchResources(initial);
+
+        updateResourcesData();
```

Comment:
In the line 82 the boolean value "true" is used, but in this case the value "undefined" will be passed. Let's setup a default value  "false" (initial = "false") or pass "false" to keep types consistency.

Reply:
Sounds good. Thank for good comments @YevheniiMaltsev